File: src/core/reportDiff.ts
// decision: split out of report.ts (base/head diffing and its markdown rendering) so the
// aggregate-scoring, diff, and human-report domains each live in their own file — see the
// coherence note in report.ts
import type { FileSummary } from './report';

export interface DiffEntry {
    filePath: string;
    baseScore: number | null;
    headScore: number;
    delta: number;
    status: 'new' | 'improved' | 'worsened' | 'unchanged';
}

// decision: keyed by filePath so callers can pass base/head summaries independently sized -
// a file present only in head (baseScore null) is reported as 'new' rather than dropped
export function diffSummaries(base: FileSummary[], head: FileSummary[]): DiffEntry[] {
    const baseByPath = new Map(base.map((f) => [f.filePath, f.score]));

    return head.map((file) => {
        const baseScore = baseByPath.get(file.filePath) ?? null;
        const delta = baseScore === null ? file.score : file.score - baseScore;
        const status: DiffEntry['status'] =
            baseScore === null ? 'new' : delta < 0 ? 'improved' : delta > 0 ? 'worsened' : 'unchanged';

        return { filePath: file.filePath, baseScore, headScore: file.score, delta, status };
    });
}

const STATUS_ICON: Record<DiffEntry['status'], string> = {
    new: '🆕',
    improved: '🟢',
    worsened: '🔴',
    unchanged: '⚪'
};

const EMPTY_STATUS_COUNTS: Record<DiffEntry['status'], number> = { new: 0, improved: 0, worsened: 0, unchanged: 0 };

export function renderDiffMarkdown(entries: DiffEntry[], baseRef: string): string {
    const lines: string[] = [];

    lines.push(`# Energy State Diff vs \`${baseRef}\``);
    lines.push('');
    lines.push('| File | Base | Head | Δ | Status |');
    lines.push('| --- | --- | --- | --- | --- |');

    for (const entry of entries) {
        const base = entry.baseScore === null ? '—' : String(entry.baseScore);
        const delta = entry.baseScore === null ? '—' : entry.delta > 0 ? `+${entry.delta}` : String(entry.delta);
        lines.push(
            `| ${entry.filePath} | ${base} | ${entry.headScore} | ${delta} | ${STATUS_ICON[entry.status]} ${entry.status} |`
        );
    }

    // decision: tallies every status in one pass over a Record rather than three separate
    // `entries.filter(e => e.status === '...').length` calls — besides being cheaper, it avoids
    // repeating e.status against string literals, which the primitive-obsession detector reads
    // as stringly-typed control flow even though DiffEntry['status'] is already a literal union
    const statusCounts = entries.reduce(
        (counts, entry) => {
            counts[entry.status] += 1;
            return counts;
        },
        { ...EMPTY_STATUS_COUNTS }
    );

    lines.push('');
    lines.push(
        `_${entries.length} file${entries.length === 1 ? '' : 's'} changed, ${statusCounts.worsened} worsened, ${statusCounts.improved} improved, ${statusCounts.new} new._`
    );

    return lines.join('\n');
}

File: src/Core/ReportDiff.fs
module Energy.Core.ReportDiff

open Energy.Core.Report

type DiffStatus =
    | New
    | Improved
    | Worsened
    | Unchanged

type DiffEntry =
    { FilePath: string
      BaseScore: int option
      HeadScore: int
      Delta: int
      Status: DiffStatus }

let diffSummaries (baseSummaries: FileSummary list) (headSummaries: FileSummary list) =
    let scores =
        baseSummaries |> List.map (fun file -> file.FilePath, file.Score) |> Map.ofList

    headSummaries
    |> List.map (fun file ->
        let baseScore = Map.tryFind file.FilePath scores

        let delta =
            baseScore
            |> Option.map (fun value -> file.Score - value)
            |> Option.defaultValue file.Score

        let status =
            match baseScore with
            | None -> New
            | Some _ when delta < 0 -> Improved
            | Some _ when delta > 0 -> Worsened
            | Some _ -> Unchanged

        { FilePath = file.FilePath
          BaseScore = baseScore
          HeadScore = file.Score
          Delta = delta
          Status = status })

let private icon =
    function
    | New -> "🆕"
    | Improved -> "🟢"
    | Worsened -> "🔴"
    | Unchanged -> "⚪"

let private name =
    function
    | New -> "new"
    | Improved -> "improved"
    | Worsened -> "worsened"
    | Unchanged -> "unchanged"

let renderDiffMarkdown (entries: DiffEntry list) (baseRef: string) =
    let rows =
        entries
        |> List.map (fun entry ->
            let baseScore =
                entry.BaseScore |> Option.map (string<int>) |> Option.defaultValue "—"

            let delta =
                if entry.BaseScore.IsNone then "—"
                elif entry.Delta > 0 then "+" + string<int> entry.Delta
                else string<int> entry.Delta

            sprintf
                "| %s | %s | %d | %s | %s %s |"
                entry.FilePath
                baseScore
                entry.HeadScore
                delta
                (icon entry.Status)
                (name entry.Status))

    let worsened =
        entries |> List.filter (fun entry -> entry.Status = Worsened) |> List.length

    let improved =
        entries |> List.filter (fun entry -> entry.Status = Improved) |> List.length

    let newFiles =
        entries |> List.filter (fun entry -> entry.Status = New) |> List.length

    let suffix = if entries.Length = 1 then "" else "s"

    [ sprintf "# Energy State Diff vs `%s`" baseRef
      ""
      "| File | Base | Head | Δ | Status |"
      "| --- | --- | --- | --- | --- |" ]
    @ rows
    @ [ ""
        sprintf
            "_%d file%s changed, %d worsened, %d improved, %d new._"
            entries.Length
            suffix
            worsened
            improved
            newFiles ]
    |> String.concat "\n"
